# Laboratorio 1: Exploración y Visualización de Datos
Fecha: Marzo 2024

Integrantes: **Vicente Leyton C. - Andrés Salazar C.**


## Declaración de compromiso ético

Nosotros/as **Vicente Leyton C. - Andrés Salazar C.**, declaramos que realizamos de manera grupal los pasos de la presente actividad. También declaramos no incurrir en copia, ni compartir nuestras respuestas con otras personas ni con otros grupos. Por lo que, ratificamos que las respuestas son de nuestra propia confección y reflejan nuestro propio conocimiento.

## Instrucciones

1.  Trabajen en equipos de dos personas. Salvo excepciones, no se corregirá entregas con menos de dos integrantes.

2.  Modifique este archivo `.ipynb` agregando sus respuestas donde corresponda. Puede ocupar Jupyter notebook en su computador o usar Google Colab como alternativa online.

3.  Para cada pregunta **incluya el código fuente que utilizó para llegar a su respuesta. Respuestas sin código no recibirán puntaje.**.

4.  El formato de entrega para esta actividad es un archivo **html**. **Genere un archivo HTML usando Jupyter** y súbalo a U-Cursos. Basta con que un/a integrante haga la entrega. Si ambos/as hacen una entrega en U-Cursos, se revisará cualquiera de éstas.

## Laboratorio

Ahora vamos a jugar con los datos!

En este laboratorio trabajaremos con un dataset del tema favorito de todxs... el anime. El dataset fue obtenido de "https://www.kaggle.com/datasets/azathoth42/myanimelist" y consta de informacion sobre animes de la pagina MyAnimeList (MAL)

Este set de datos consiste de dos archivos. "anime.csv" contiene 6 atributos: "title", "type", "source", "episodes", "airing" y "studio", que corresponden a informacion sobre el anime, y "anime-stats.csv" contiene 4 atributos: "title", "score", "members" (numero de usuarios que tienen el anime en su lista) y "favorites", que corresponden a estadisticas en MyAnimeList sobre la popularidad de animes.

Para cargar los datos desde tu computador, usa:

In [10]:
import pandas as pd
import matplotlib.pyplot as plt

In [ ]:
#los archivos deben estar en la carpeta del notebook
df_anime = pd.read_csv('anime.csv', encoding="UTF-8")
df_stats = pd.read_csv('anime-stats.csv', encoding="UTF-8")

También puedes cargarlos remotamente de un repositorio usando el link a los archivos:


In [11]:
df_anime = pd.read_csv('https://raw.githubusercontent.com/bacchus00/cositas/main/anime.csv', encoding="UTF-8")
df_stats = pd.read_csv('https://raw.githubusercontent.com/bacchus00/cositas/main/anime-stats.csv', encoding="UTF-8")

### Exploración básica


1.  ¿Cuáles son las dimensiones de ambos dataframes (filas, columnas)?


In [12]:
anime_dimentions = df_anime.shape
stats_dimentions = df_stats.shape
print(f"anime dimentions: {anime_dimentions} \nstats dimentions: {stats_dimentions}")

anime dimentions: (14478, 6) 
stats dimentions: (14478, 4)


Las dimensiones para el dataframe de anime son (14478, 6) y para el de stats son (14478,4)

**Fuente**

metodo shape: https://www.delftstack.com/es/howto/python-pandas/how-to-get-the-row-count-of-a-pandas-dataframe/#el-m%c3%a9todo-shape-para-obtener-el-n%c3%bamero-de-filas-de-dataframe

2.  ¿Existen filas duplicadas en alguno de los dataframes?


In [13]:
print(df_anime.duplicated().any())
print(df_stats.duplicated().any())

False
False


Como podemos ver, no existen filas duplicadas para ninguno de los dataframes

https://stackoverflow.com/questions/60753830/find-duplicate-rows-from-df-python#:~:text=duplicated%20actually%20returns%20a%20Series%20containing%20boolean%20values,if%20there%20is%20any%20duplicate%20in%20your%20DataFrame.

3.  ¿Hay filas con valores nulos en alguno de los dataframes?

In [14]:
print(df_anime.isnull().any().any())
print(df_stats.isnull().any().any())

True
False


Si, el dataframe de anime contiene filas con valores nulos, pero el de stats no.

**Fuentes**

* https://www.analyticslane.com/2021/08/12/pandas-contar-los-valores-nulos-en-dataframe/

* https://es.stackoverflow.com/questions/221050/conocer-si-una-fila-de-un-dataframe-tiene-un-valor-nan-pandas#:~:text=Si%20yo%20quiero%20saber%20si%20hay%20alg%C3%BAn%20valor,quiero%20saber%20valor%20a%20valor%2C%20a%20%3D%20df.isnull%28%29

### Análisis

Ahora para hacer el análisis más sencillo juntaremos ambas tablas  usando como llave el nombre del anime (title).

In [15]:
df = df_anime.merge(df_stats, on='title')
df.sample(5)

title   type       source  \
2871   JoJo no Kimyou na Bouken: Diamond wa Kudakenai     TV        Manga   
11986                     Toaru Hikuushi e no Tsuioku  Movie  Light novel   
13045                             Zo Zo Zo Zombie-kun    ONA        Manga   
11811             Detective Conan OVA 02: 16 Suspects    OVA        Manga   
10035                    Gokujou!! Mecha Mote Iinchou     TV        Manga   

       episodes  airing             studio  score  members  favorites  
2871         39   False   David Production   8.63   181435       7188  
11986         1   False           Madhouse   7.75    36647        120  
13045         2   False                NaN   4.42       69          0  
11811         1   False  TMS Entertainment   7.42    10410          9  
10035        51   False          SynergySP   6.68     4996         17

1. Lista los animes que tienen una cantidad de episodios mayor o igual a 1000.

In [16]:
df_filtrado = df[df["episodes"] >= 1000]
print(df_filtrado)

                                       title type    source  episodes  airing  \
3566                   Sekai Monoshiri Ryoko   TV   Unknown      1006   False   
3815                  Kirin Monoshiri Yakata   TV   Unknown      1565   False   
3816      Manga Nippon Mukashibanashi (1976)   TV   Unknown      1471   False   
6231   Monoshiri Daigaku: Ashita no Calendar   TV  Original      1274   False   
7765                        Hoka Hoka Kazoku   TV   Unknown      1428   False   
11217                        Doraemon (1979)   TV     Manga      1787   False   
11629                             Oyako Club   TV  Original      1818   False   
13573               Kirin Ashita no Calendar   TV   Unknown      1306   False   

                  studio  score  members  favorites  
3566                 NaN   6.09      244          0  
3815                 NaN   5.98      198          1  
3816           Group TAC   6.37      619          2  
6231                 NaN   6.42      211          0  


https://www.delftstack.com/es/howto/python-pandas/how-to-filter-dataframe-rows-based-on-column-values-in-pandas/

2. Encuentra los 5 animes con mas favoritos según el dataset ¿Te hace sentido que estos sean los animes con mas favoritos?


In [17]:
df.sort_values(by=["favorites"], ascending=False).head(5)

title type        source  episodes  airing  \
2555  Fullmetal Alchemist: Brotherhood   TV         Manga        64   False   
3000                       Steins;Gate   TV  Visual novel        24   False   
7914                        Death Note   TV         Manga        37   False   
34                           One Piece   TV         Manga         0    True   
418             Hunter x Hunter (2011)   TV         Manga       148   False   

              studio  score  members  favorites  
2555           Bones   9.25  1199091     106895  
3000       White Fox   9.14  1010330      92423  
7914        Madhouse   8.67  1456378      88696  
34    Toei Animation   8.54   720133      69760  
418         Madhouse   9.11   720920      64375

https://stackoverflow.com/questions/47462690/how-to-get-top-5-values-from-pandas-dataframe

3.  Lista todos los studios (sin repetirse y sin incluir valores nulos) que tienen animes en entre los 20 con mejor score.

In [18]:
top_20 = df.dropna().sort_values(by=["score"], ascending=False).head(20)
studios = top_20.filter(["studio"]).drop_duplicates()
print(studios)

                      studio
14405    Collaboration Works
2555                   Bones
1028        CoMix Wave Films
1252   Bandai Namco Pictures
6178               White Fox
418                 Madhouse
8993                 Sunrise
12726                Artland
1307                   Shaft
6573         Kyoto Animation
6512          Production I.G
7106           Studio Ghibli


Dropna (Eliminar filas con nulos) https://www.freecodecamp.org/espanol/news/limpieza-de-datos-en-pandas-explicado-con-ejemplos/#:~:text=Dropna,que%20para%20unas%20cuantas%20variables.

drop_duplicates (Eliminar filas duplicadas) https://www.delftstack.com/es/howto/python-pandas/drop-duplicates-pandas/

4. ¿Hay algún studio que tenga más de un anime entre los 20 con mejor score? (sin repeticiones y sin incluir nulos)


In [19]:
top_20 = df.dropna().sort_values(by=["score"], ascending=False).head(20)
studios = top_20.filter(["studio"])
studios_duplicados = studios[studios.duplicated()].drop_duplicates()
print(studios_duplicados) 

                      studio
3000               White Fox
5784                 Sunrise
10880  Bandai Namco Pictures
6616                   Shaft
2783         Kyoto Animation


Si, estos son los estudios que tienen más de un anime entre los 20 con mejor score

Y eso es todo por hoy :)


![](https://i.imgur.com/ghKLs78.jpeg)